consensus/src/common.rs

```diff
@@ -1,40 +1,57 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0
 
+//! Implements common APIs for the blockchain in the context of consensus.
+
 use hex::ToHex;
-use mina_crypto::hash::*;
-use mina_rs_base::types::*;
-use thiserror::Error;
+use mina_crypto::hash::{Hashable, StateHash};
+use mina_rs_base::consensus_state::ConsensusState;
+use mina_rs_base::global_slot::GlobalSlot;
+use mina_rs_base::protocol_state::{Header, ProtocolState};
 
-pub struct ProtocolStateChain(pub Vec<ProtocolState>);
+use crate::density::{relative_min_window_density, ConsensusConstants};
+use crate::error::ConsensusError;
 
-#[derive(Error, Debug, PartialEq)]
-pub enum ChainError {
-    #[error("header must have height 1 greater than top")]
-    InvalidHeight,
-}
+#[derive(Clone, Debug, Default)] // FIXME: remove clone here.
+                                 // TODO: replace vec element with ExternalTransition
+pub struct ProtocolStateChain(pub Vec<ProtocolState>);
 
 pub trait Chain<T>
 where
     T: Header,
 {
-    fn push(&mut self, new: T) -> Result<(), ChainError>;
+    fn push(&mut self, new: T) -> Result<(), ConsensusError>;
+    /// This function returns the last block of a given chain.
+    /// The input is a chain C and the output is last block of C
+    /// (i.e. the block with greatest height).
     fn top(&self) -> Option<&T>;
+    /// he function returns the consensus state of a block or chain.
+    /// The input is a block or chain X and the output is the consensus state.
     fn consensus_state(&self) -> Option<&ConsensusState>;
+    /// The function returns the global slot number of a chain or block.
+    /// The input X is either a chain or block and the output is the global slot number.
     fn global_slot(&self) -> Option<&GlobalSlot>;
+    // The function computes the epoch slot number of a block.
```

Comment:
Once #105 is merged these will have to be docstring comments `\\\` and all methods must be commented

Reply:
done